llvm/lib/Target/AMDGPU/GCNHazardRecognizer.cpp

```diff
@@ -1417,9 +1417,8 @@ static bool shouldRunLdsBranchVmemWARHazardFixup(const MachineFunction &MF,
   bool HasVmem = false;
   for (auto &MBB : MF) {
     for (auto &MI : MBB) {
-      HasLds |= SIInstrInfo::isDS(MI);
-      HasVmem |= (SIInstrInfo::isVMEM(MI) && !SIInstrInfo::isFLAT(MI)) ||
-                 SIInstrInfo::isSegmentSpecificFLAT(MI);
+      HasLds |= SIInstrInfo::isDS(MI) || SIInstrInfo::isLDSDMA(MI);
+      HasVmem |= SIInstrInfo::isVMEM(MI) && !SIInstrInfo::isLDSDMA(MI);
```

Comment:
Without having a deep understanding of the hazard, it is not clear to me why you would exclude isLDSDMA here. I.e. why shouldn't an instruction like GLOBAL_LOAD_LDS_DWORD satisfy both the HasLds and HasVmem conditions?

Reply:
> Without passing through VGPRs would mean that no vector memory is touched

No, a buffer load is still reading from "vector memory" even if it does not touch VGPRs.

Reply:
Hm ok. But previously, we had `(SIInstrInfo::isVMEM(MI) && !SIInstrInfo::isFLAT(MI)) || SIInstrInfo::isSegmentSpecificFLAT(MI)`. And isn't that basically meaning that we want to have FLAT instructions as long as they don't do LDS DMA?

Reply:
> I think the intent here was to check for GLOBAL_* or SCRATCH_* instructions, excluding FLAT_* instructions (but all 3 sets are FLAT encoded instructions)

Agreed, that matches the GFX9/GFX10 docs that I have found. Only MUBUF, GLOBAL_* and SCRATCH_* instructions support load-to-LDS.

Reply:
So what we actually want for `HasVmem` is just plain `SIInstrInfo::isVMEM(MI)`?